### DISCUSSION TITLE
Help when trying to upload a PDF to a digital listing on Etsy

### DISCUSSION THREAD

**Discussion by DevDenzo** [etsy/open-api#1056] [Q&A]
I am trying to upload a PDF onto a listing that I have in draft status. I have looked at the etsy documentation and believe this is the correct endpoint to use:

`https://openapi.etsy.com/v3/application/shops/{shop_id}/listings/{self.listing_id}/files`

I have chosen to go with Python and My function is:

```
def addFileToDraftListing(self, shop_id):
        
        headers = {
            "Authorization": "Bearer " + self.data["etsy_token"]["access_token"],
            "x-api-key": self.data["etsy_keystring"]
        }

        refresh_url = "https://api.etsy.com/v3/public/oauth/token"
        etsy_auth = OAuth2Session(self.data["etsy_keystring"], token=self.data["etsy_token"], auto_refresh_url=refresh_url)
        
        file_path = "1.pdf"
        with open(file_path, 'rb') as f:
            file_data = f.read()
            file_name = os.path.basename(file_path)

        data = {
            'name':  file_name,
        }

        files={
            'name': file_name, 
            'image': (file_path, file_data, 'image')} 

        r = etsy_auth.post(f"https://openapi.etsy.com/v3/application/shops/{shop_id}/listings/{self.listing_id}/files", headers=headers, files=files, data=data)

        print(r.text)
```

The result I am getting is:
{"error":"Either a valid listing_file_id or file must be provided."}

The etsy documentation states that the only path params required is the shop_id and listing_id which I provide. It is asking for a valid listing_file_id however I can not obtain one as no file has already been uploaded.

Can someone help me fix the issue I am facing? I will also need a similar solution for the upload of listing images so this code is very important for me to crack. The solution I have above is a mixture of all the other threads I have found however, none of the old solutions are working for me either.

**Comment by Gareth064** [comment]
I'm not familiar with your language. But is the file information being submitted as form data as per the spec?

Sent from Outlook for Android<https://aka.ms/AAb9ysg>
________________________________
From: Deniz Yukselir ***@***.***>
Sent: Friday, July 7, 2023 9:53:56 PM
To: etsy/open-api ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [etsy/open-api] Help when trying to upload a PDF to a digital listing on Etsy (Discussion #1056)


I am trying to upload a PDF onto a listing that I have in draft status. I have looked at the etsy documentation and believe this is the correct endpoint to use:

https://openapi.etsy.com/v3/application/shops/{shop_id}/listings/{self.listing_id}/files

My function is:

`
def addFileToDraftListing(self, shop_id):

    headers = {
        "Authorization": "Bearer " + self.data["etsy_token"]["access_token"],
        "x-api-key": self.data["etsy_keystring"]
    }

    refresh_url = "https://api.etsy.com/v3/public/oauth/token"
    etsy_auth = OAuth2Session(self.data["etsy_keystring"], token=self.data["etsy_token"], auto_refresh_url=refresh_url)

    file_path = "1.png"
    with open(file_path, 'rb') as f:
        file_data = f.read()
        file_name = os.path.basename(file_path)

    data = {
        'name':  file_name,
    }

    files={
        'name': file_name,
        'image': (file_path, file_data, 'image')}

    r = etsy_auth.post(f"https://openapi.etsy.com/v3/application/shops/{shop_id}/listings/{self.listing_id}/files", headers=headers, files=files, data=data)

    print(r.text)


`

The result I am getting is:
{"error":"Either a valid listing_file_id or file must be provided."}

The etsy documentation states that the only path params required is the shop_id and listing_id which I provide. It is asking for a valid listing_file_id however I can not obtain one as no file has already been uploaded.

Can someone help me fix the issue I am facing? I will also need a similar solution for the upload of listing images so this code is very important for me to crack. The solution I have above is a mixture of all the other threads I have found however, none of the old solutions are working for me either.

**Reply by DevDenzo** [reply]
I have tried that by changing the headers, however, that also isn't working:

```
        headers = {
            "Authorization": "Bearer " + self.data["etsy_token"]["access_token"],
            "x-api-key": self.data["etsy_keystring"],
            "Content-Type": "multipart/form-data",
            "Content-Disposition": "multipart/form-data",
        }
        
```
I also tried changing it where I specify the files:

`files = {"file" : ( file_name, open(file_path, 'rb'), "multipart/form-data")}`

That also didn't work.

**Comment by DevDenzo** [comment]
Found a solution thanks to this StackOverflow post:

https://stackoverflow.com/questions/76365459/etsy-api-file-upload-function-doesnt-work?rq=2
